### ISSUE TITLE
System/standard library for bignumbers

### ISSUE THREAD

**Issue by axic** [ewasm/design#189]
EVM supports 256-bit operations "natively". While that precision is not needed in a lot of cases it does come as very useful for certain operations, such as handling token balances, in Ethereum.

With wasm one would need to include a bignumber library to do that, potentially in a lot of contracts.  This suggests a good opportunity to define a standard bignumber library, which can be used by contracts, but it is not mandatory to be used.

Due to the design of "host" functions in wasm (having a separation of namespaces) it is possible to design this in a forward compatible manner, but implement it right now in the clients, such as Hera.

Lets define a new namespace `stdbn` (jk, `bignum` or `stdbignum`) with the following
- `mul256(a: i32ptr, b: i32ptr, out: i32ptr)`
- `mulmod256(a: i32ptr, b: i32ptr, mod: i32ptr, out: i32ptr)`

All of the arguments are 32-bit pointers to a memory location containing a 256-bit little endian number. The pointers can overlap or be the same.

In the future this could also be implemented as a system contract: #17.

**Comment by chfast** [comment]
I'm not sure the `std` prefix makes any difference here.

The name `mul256` is fine because it's the same for signed and unsigned number. But the second one should be rather `umulmod256`.

**Comment by axic** [comment]
I like the pedantic option.

**Comment by axic** [comment]
I think we should add all 256-bit methods to the host interface: `add`, `sub`, `mul`, `div`, `mulmod`, `cmp`. Also need to consider support for signed numbers.

We could argue that some of them are fast enough if implemented in Wasm, but the secondary goal here is to reduce code duplication and it helps in that.

Any comments?

**Comment by jakelang** [comment]
I agree with @axic, `bn` is a tad short and can also mean Barreto-Naehrig (a family of curves). Perhaps `bignum` would be better as a namespace, although `std_bn` or `stdbn` could suffice.

**Comment by jakelang** [comment]
Personally I feel like it would complicate the design for the bignum library to only contain the arithmetic methods that are not fast enough in wasm.

Reducing code duplication is definitely another big benefit of having a "complete" bignum lib. As long as we can minimize host function overhead, this is great for reducing contract bloat.

**Comment by axic** [comment]
Potentially we could make use of the references proposal for Wasm. Let assume the following example is for fixed 256-bit long bignums:

```
bignum.load(memOffset: i32) -> anyref
bignum.store(memOffset: i32, v: anyref)
bignum.mul(a: anyref, b: anyref)
bignum.add(a: anyref, b: anyref)
...
```

This assumes that opaque references (what `anyref` is) would be efficiently passed around in interpreters.

**Comment by axic** [comment]
Let's agree on this first basic API, which isn't using a bignum stack:
- `addu256(a: i32ptr, b: i32ptr, c: i32ptr)` where `c = a + b`
- `subu256(a, b, c)` where `c = a - b`
- `mulu256(a, b, c)` where `c = a * b`
- `divu256(a, b, c)` where `c = a / b`
- `mulmodu256(a, b, c, d)` where `d = (a * b) % c`
- `cmpu256(a: i32ptr, b: i32ptr) -> i32` where the returned value is less than zero if `a < b`, zero if `a == b` or larger than zero if `a > b`

@chfast @s1na what do you think?

**Comment by poemm** [comment]
How about the first argument is the output. so output position is consistent for different numbers of arguments, for example `binary_op(output,input1,input2)` and `unary_op(output, input)`?

**Comment by axic** [comment]
I thought about having output as the first operand, but wasn't sure of the merit. Some POSIX functions do that, but every and I do find some of them confusing.

What would be the benefit we gain? Note, I expect this to change during the upcoming months. We also need to make a decision whether to replace (or add) a stack based version.

**Comment by poemm** [comment]
> What would be the benefit we gain?

Convention. Gnu Multiple Precision and lots of crypto pass output first. But it doesn't matter.

> We also need to make a decision whether to replace (or add) a stack based version.

Not sure which algorithms benefit from the stack based version, other than pathological benchmarks. But it may be simple enough to implement -- `bin_op256_stack()` can just call `bin_op256(stack_ptr-32, stack_ptr-32, stack_ptr)` and update the stack pointer.

**Comment by jakelang** [comment]
I agree with @axic here.
Although it's somewhat common in POSIX and other legacy conventions to use `op(dst, src)`,
It is far more linguistically intuitive writing `op(src, dst)` when lacking the muscle memory for the former.

**Comment by chfast** [comment]
> I agree with @axic here.
> Although it's somewhat common in POSIX and other legacy conventions to use `op(dst, src)`,
> It is far more linguistically intuitive writing `op(src, dst)` when lacking the muscle memory for the former.

This actually go against many other conventions, e.g. Go and other OOP designs (to some distinct): https://golang.org/pkg/math/big/, RISC assembly, C calling convention.

Also `a += 1` becomes `add(a, 1, a)` instead of `add(a, a, 1)`.

**Comment by axic** [comment]
@chfast can you clarify which convention you referred to or which one you prefer?

**Comment by chfast** [comment]
I prefer the outputs to be the first arguments.

**Comment by s1na** [comment]
A couple more discussion points:

- What about returning a carry where necessary, e.g. `add256`? @cdetrio is doing it [here](https://github.com/cdetrio/scout.ts/blob/358b99ea1b35edfde885a82ee1afd84e6e04b840/src/lib.ts#L200) to maintain compatibility with websnark (the result is also in LE).
- Whether montgomery multiplication (https://github.com/cdetrio/scout.ts/blob/358b99ea1b35edfde885a82ee1afd84e6e04b840/src/lib.ts#L133) should be part of the api. Probably needs benchmarks.
- How to handle arbitrarily large numbers

**Comment by poemm** [comment]
> What about returning a carry where necessary, e.g. add256?

Makes sense to return a carry. It may also make sense to have an extra argument for carry.

> Whether montgomery multiplication should be part of the api.

It should. Modular multiplication is a bottleneck for elliptic curve crypto. `MontgomeryMultiplication256` is very popular for this. We may also want `MontgomeryReduction256` to transform to/from Montgomery form, and to speed up squaring. We may eventually consider `BarrettMultiplication256` which is popular too. A problem is metering - some of these are faster with non-constant runtime, so we may have to meter conservatively, or break the algorithms into constant-runtime chunks, and meter each chunk individually.

> How to handle arbitrarily large numbers

@chfast wisely said that anything bigger than `mul256` will have register pressure and a slowdown. And that many operations can be built by composition of smaller algorithms. For example, we can build `mul768` from `mul256x256->512`s. For this reason, we may eventually want `mul64x64->128` and `mul128x128->256` as building blocks, but this is less important for now.

**Comment by axic** [comment]
> What about returning a carry where necessary, e.g. add256?

We kind of agreed to have another set of methods with carry.

**Comment by s1na** [comment]
Proposing this slightly modified API:

- `add(width: i32, a: i32ptr, b: i32ptr, c: i32ptr): bool` where `c = a + b`, width is for both `a` and `b`, and return value indicates carry
- `sub(width: i32, a, b, c): bool` where `c = a - b`, and return value indicates carry (`c < 0`)
- `mul(width: i32, a, b, c): bool` where `c = a * b`, and return value indicates carry
- `div(width: i32, a, b, q, r): void` where `q = a / b` and `r = a % b`
- `mulmod(a, b, c, d): bool` where `d = (a * b) % c` and return value indicates `(a * b) > c` (not sure if width is needed here?)
- `cmp(width: i32, a: i32ptr, b: i32ptr): i32` where the returned value is less than zero if a < b, zero if a == b or larger than zero if a > b

Note: these changes only make sense if the overhead of additional parameters and return values is negligible.

**Comment by chfast** [comment]
Carry for `mul` is not practical. It's a bit of trouble to get it, but is not useful in higher precision multiplication.

**Comment by s1na** [comment]
@chfast Thanks for the input. Removed carry for `mul` and `mulmod`.
